### DISCUSSION TITLE
How to send hex to device?

### DISCUSSION THREAD

**Discussion by preflower** [kenjdavidson/react-native-bluetooth-classic#85] [General]
![image](https://user-images.githubusercontent.com/30513719/84018413-9ce8d980-a9b2-11ea-8bc5-81e9117b9694.png)
i find write param type is String, but i need to send hex to device? how to do it?

**Comment by preflower** [comment]
![image](https://user-images.githubusercontent.com/30513719/84023622-ffde6e80-a9ba-11ea-9ddc-bdbc35b8090e.png)
i find write that support hex, and i send Buffer but not response

**Comment by kenjdavidson** [comment]
Yes. At this point in time the response is solely string. There was an issue opened early in the project where someone returned a byte array. You'd need to do the same thing as that yourself.

I have the plan to allow writing custom parsers in the bootstrapping, but don't have the time to complete it at this point.

Since the result is a string, you should be able to convert it to hex representation yourself.

**Comment by preflower** [comment]
> Yes. At this point in time the response is solely string. There was an issue opened early in the project where someone returned a byte array. You'd need to do the same thing as that yourself.
> 
> I have the plan to allow writing custom parsers in the bootstrapping, but don't have the time to complete it at this point.
> 
> Since the result is a string, you should be able to convert it to hex representation yourself.

i use method of Self-Transmitting and Self-Receiving test; but i find the data received is inconsistent with the data sent;  the data is correct in writeToDevice function, but onReceivedData i get a incorrect data；can you help me check it?

**Comment by preflower** [comment]
> > Yes. At this point in time the response is solely string. There was an issue opened early in the project where someone returned a byte array. You'd need to do the same thing as that yourself.
> > I have the plan to allow writing custom parsers in the bootstrapping, but don't have the time to complete it at this point.
> > Since the result is a string, you should be able to convert it to hex representation yourself.
> 
> i use method of Self-Transmitting and Self-Receiving test; but i find the data received is inconsistent with the data sent; the data is correct in writeToDevice function, but onReceivedData i get a incorrect data；can you help me check it?

send file:
```
let buf = new Buffer('3aa30000000000010002800183a8', 'hex');
RNBluetoothClassic.write(buf)
```
RNBluetoothClassicModule.java
```
@Override
  public void onReceivedData(BluetoothDevice device, byte[] data) {
    String msg = String.format("Received %d bytes from device %s", data.length, device.getName());
    Log.d(TAG, msg);

    mBuffer.append(new String(data, mCharset));

    if (!mReadObserving.get()) {
      Log.d(TAG, "No BTEvent.READ listeners are registered, skipping handling of the event");
      return;
    }

    char[] hexArray = "0123456789ABCDEF".toCharArray();
    char[] hexChars = new char[data.length * 2];
    for (int j = 0; j < data.length; j++) {
        int v = data[j] & 0xFF;
        hexChars[j * 2] = hexArray[v >>> 4];
        hexChars[j * 2 + 1] = hexArray[v & 0x0F];
    }

    String message = new String(hexChars);
    Log.d(TAG, "Check Data:" + message);
    if (message != null) {
      BluetoothMessage bluetoothMessage
        = new BluetoothMessage<>(new NativeDevice(mBluetoothService.connectedDevice()).map(), message);
      sendEvent(BluetoothEvent.READ.code, bluetoothMessage.asMap());
    }
  }
```

**Comment by preflower** [comment]
![image](https://user-images.githubusercontent.com/30513719/84239924-d3982e80-ab2f-11ea-8ba9-859ec6f2083b.png)
i find reason, bytes[] change UTF-8 type in write function, and this operation will result in the hex code inconsistency required for the machine. so i edit write and it's right now

**Reply by derb21** [reply]
What was your fix?

**Comment by kenjdavidson** [comment]
Thanks for looking into it, that would definitely cause you issues attempting to send HEX. 

I don't know how that will affect users already using as it is (as mentioned somewhere, most of the Android stuff I just pulled over from react-native-bluetooth-serial).  I've slowly been re-writing to include more configuration, like adding middleware/handlers to the on and receive that can be applied to the device during the `connect` request.

I just haven't had time to finish.  also to get the same functionality on Apple, I need to rewrite a bunch of that so that its either All Swift or All Objective C. 

Hopefully you can work off your own changes for the time being, but I will definitely leave this open so I have another instance of configured data.

**Comment by derb21** [comment]
Did this get resolved? I am trying to also send a byte array image to my printer but getting base64 random characters. Any fix for this issue?

**Comment by derb21** [comment]
{"data": [0, 192, 0, 120, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 3, 192, 0, 3, 224, 0, 3, 240, 0, 1, 252, 0, 1, 254, 0, 0, 127, 0, 0, 63, 0, 0, 31, 0, 0, 7, 0, 0, 3, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 192, 0, 1, 224, 0, 1, 240, 0, 0, 248, 0, 0, 252, 0, 0, 126, 0, 0, 63, 0, 0, 31, 0, 0, 15, 128, 0, 7, 192, 0, 3, 240, 0, 1, 248, 0, 0, 252, 0, 0, 255, 0, 0, 255, 128, 0, 127, 192, 0, 63, 240, 0, 14, 248, 0, 7, 124, 0, 3, 191, 0, 1, 207, 128, 0, 231, 192, 0, 113, 240, 0, 56, 248, 0, 28, 126, 0, 14, 31, 0, 7, 15, 0, 3, 131, 0, 1, 193, 0, 0, 224, 0, 0, 112, 0, 0, 56, 0, 0, 28, 0, 0, 14, 0, 0, 14, 0, 0, 7, 0, 0, 3, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 252, 0, 1, 255, 0, 1, 255, 0, 1, 199, 0, 1, 128, 0, 1, 192, 0, 1, 192, 0, 1, 192, 0, 0, 224, 0, 0, 224, 0, 0, 112, 0, 0, 112, 0, 0, 56, 0, 0, 60, 0, 0, 28, 0, 0, 30, 0, 0, 14, 0, 0, 7, 0, 0, 7, 0, 0, 3, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 63, 0, 255, 255, 3, 255, 255, 7, 255, 255, 15, 255, 192, 15, 0, 0, 14, 0, 0, 28, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 192, 0, 0, 224, 0, 0, 240, 0, 0, 248, 0, 0, 252, 0, 0, 254, 0, 0, 127, 0, 0, 63, 128, 0, 31, 224, 0, 7, 240, 0, 3, 248, 0, 1, 252, 0, 0, 254, 0, 0, 127, 0, 0, 63, 128, 0, 31, 192, 0, 15, 224, 0, 7, 240, 0, 3, 248, 0, 1, 252, 128, 0, 254, 192, 0, 127, 240, 0, 63, 248, 0, 31, 60, 0, 15, 31, 0, 7, 15, 128, 3, 3, 192, 1, 1, 224, 0, 0, 248, 0, 128, 60, 0, 192, 30, 0, 224, 15, 0, 112, 7, 128, 48, 3, 224, 56, 0, 240, 28, 0, 120, 14, 0, 60, 7, 0, 30, 3, 0, 15, 3, 128, 7, 1, 192, 1, 0, 224, 0, 0, 96, 0, 0, 112, 0, 0, 56, 0, 0, 28, 0, 0, 12, 0, 0, 14, 0, 0, 7, 0, 0, 3, 0, 128, 3, 128, 224, 1, 192, 252, 0, 224, 255, 0, 224, 31, 224, 112, 7, 248, 56, 1, 255, 56, 0, 63, 220, 0, 15, 252, 0, 3, 254, 0, 0, 127, 0, 0, 31, 0, 0, 7, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 128, 0, 0, 128, 0, 0, 192, 0, 0, 192, 0, 0, 224, 0, 0, 240, 0, 0, 112, 0, 0, 56, 0, 0, 60, 0, 0, 28, 0, 0, 14, 0, 0, 14, 0, 0, 7, 0, 0, 3, 128, 0, 3, 128, 0, 1, 192, 0, 1, 192, 0, 0, 224, 0, 0, 96, 0, 0, 112, 0, 0, 48, 0, 0, 56, 0, 0, 24, 0, 0, 28, 0, 0, 12, 0, 0, 14, 0, 0, 6, 0, 0, 7, 0, 0, 3, 0, 0, 3, 128, 0, 1, 128, 0, 1, 128, 0, 1, 192, 0, 0, 192, 0, 0, 224, 0, 0, 96, 0, 0, 96, 0, 0, 112, 0, 0, 48, 0, 0, 48, 0, 0, 56, 0, 0, 24, 0, 0, 24, 0, 0, 24, 0, 0, 28, 0, 0, 12, 192, 0, 12, 224, 0, 12, 248, 0, 12, 254, 0, 14, 63, 128, 14, 31, 224, 14, 7, 248, 14, 1, 255, 12, 0, 127, 252, 0, 31, 252, 0, 3, 248, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128, 0, 0, 192, 0, 0, 224, 0, 0, 112, 0, 0, 184, 0, 0, 220, 0, 0, 238, 0, 0, 119, 0, 0, 127, 128, 0, 59, 192, 0, 29, 192, 0, 14, 224, 0, 7, 112, 0, 7, 184, 0, 3, 188, 0, 1, 220, 0, 0, 238, 0, 0, 103, 0, 128, 115, 128, 192, 59, 128, 224, 25, 192, 112, 28, 224, 56, 14, 112, 28, 6, 112, 14, 7, 56, 7, 3, 156, 3, 131, 158, 1, 193, 206, 0, 224, 199, 0, 112, 231, 0, 56, 99, 0, 28, 113, 0, 14, 49, 0, 7, 56, 0, 3, 184, 0, 1, 216, 0, 0, 248, 0, 0, 124, 0, 0, 60, 128, 0, 28, 224, 0, 14, 248, 0, 15, 255, 0, 15, 255, 224, 29, 239, 255, 249, 227, 255, 248, 112, 127, 240, 48, 0, 0, 56, 0, 0, 28, 0, 0, 28, 0, 0, 14, 0, 0, 14, 0, 0, 7, 0, 0, 7, 0, 0, 3, 128, 0, 3, 128, 0, 1, 192, 0, 1, 192, 0, 0, 224, 0, 0, 224, 0, 0, 112, 0, 0, 112, 0, 0, 48, 0, 0, 56, 0, 0, 56, 0, 0, 24, 0, 0, 28, 0, 0, 28, 0, 0, 12, 0, 0, 12, 0, 0, 14, 0, 0, 14, 0, 0, 6, 0, 0, 6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128, 0, 0, 128, 0, 0, 192, 0, 0, 224, 0, 0, 224, 0, 0, 112, 0, 0, 120, 0, 0, 56, 0, 0, 60, 0, 0, 28, 0, 0, 14, 0, 0, 15, 0, 0, 7, 0, 0, 135, 128, 0, 195, 128, 0, 227, 192, 0, 241, 192, 0, 121, 224, 0, 56, 224, 0, 28, 240, 0, 14, 112, 0, 15, 112, 0, 7, 240, 0, 3, 240, 0, 1, 240, 0, 0, 240, 0, 0, 112, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], "type": "Buffer"}
This is what I am trying to send.

**Comment by kenjdavidson** [comment]
They is documentation and one of the most recent forks changed some things to use hex arrays instead of strings. You can check them out.

I have to get the hexarray into the library by default a more and more twists are coming for it.

If you can test the last fork, and it works, I will pull those changes.

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Reply by derb21** [reply]
Whats the most recent fork? do you have a link? and documentation for hex array link? Thanks for the reply

**Reply by kenjdavidson** [reply]
I believe it's this one https://github.com/zyxit2/react-native-bluetooth-classic
Just take a look through the latest recent forks and check out the commits ahead, you should see something to do with `ByteArrayConnectionImpl`

**Reply by derb21** [reply]
Okay, I found the fork and it looks like it's only for android? Is this update only for android?

**Reply by kenjdavidson** [reply]
The whole `ByteArrayConnectionImpl` (was done on my side) just to show how to extend and create your own `Connection` types for yourself.  Using the overriding of the React Native Autoloading features you can design any kind of functionality you want for both Android or IOS - I never took it so far as to add into IOS (mainly because I dislike (hate is a little too strong) Apple and Apple development).

Feel free to take the concept of `ByteArrayConnectionImpl` and extend it into IOS, I will gladly accept a PR that has `ByteArrayConnectionImpl` working on both Android and IOS with documentation.

**Reply by derb21** [reply]
Okay sounds good. Thanks for helping me out. I tried to write a hex array using the latest fork. But It's giving me the same response.(printed ticket)  Do I just use the same writeToDevice()?     

let write = await RNBluetoothClassic.writeToDevice(connected.address, byteArr, "hex");

**Reply by kenjdavidson** [reply]
I haven't really looked into the details of what he's done in the fork, so I can't be sure.  When I was playing around with my own `ByteArrayConnectionImpl` you would use the same call to write.

https://kenjdavidson.com/react-native-bluetooth-classic/guides/hex-encoding-decoding/

Here is the Guide on how I envisioned it working, please let me know how close it is to what you need.

Sorry I can't give you any more - work is so crazy that I haven't really been able to finish what I started for it.

**Reply by kenjdavidson** [reply]
When I get around to it, I plan on adding `ByteArrayConnectionImpl` to the standard `RNBluetoothClassicPackage` (since it's being asked about a whole bunch).

**Reply by kenjdavidson** [reply]
The problem is, I don't have access to an MFi compliant device to test with on IOS.

**Reply by derb21** [reply]
Possible Unhandled Promise Rejection (id: 5):
TypeError: connected.write is not a function. (In 'connected.write(signaturePacked, "hex")', 'connected.write' is undefined)

I have to do writeToDevice, can't do write like the documentation suggests to do. Is this a bug?

**Reply by derb21** [reply]
No worries, I understand. This is the best documentation I have found regarding Bluetooth serial for RN so I really appreciate it.

**Comment by kenjdavidson** [comment]
Ya. I think I fixed it in the last release.

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Comment by kenjdavidson** [comment]
Lol. It drove me crazy so I needed to do it. Good luck.

Get Outlook for Android<https://aka.ms/ghei36>
________________________________

**Comment by derb21** [comment]
@kenjdavidson I am working on allowing to write a byte array to my brother printer but all I get is a bunch of random characters. My company is willing to pay someone to add this functionality, they wanted me to ask you if you would be wiling to update this for IOS and Android and what it would cost if you had the time to do this.

**Comment by kenjdavidson** [comment]
Hey man, at this point I don't really have any time to commit to something like this (my day job is on a big time crunch).  I have no problem guiding you or looking at code if you have working example(s) though.  First off:

- Do you have the Apple MFi protocols for the printer(s) in question?  Without these you're out of luck with regards to getting them work. 
- Have you followed some of the documentation / forks that are available?
- Do you have the the Printer API available, there are a ton of Brother printers (are they all the same API)?

A number of users have forked this repo and updated the `DeviceConnectionImpl` class to allow for byte array's to be transferred.  I've also included the `ByteArrayDeviceConnectionImpl` which can be tested - although sadly I don't have any bluetooth devices to test with at this point in time (which is why I haven't actually made it a standard connection type yet.
